File: pkg/clients/product_client.go
package clients

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/aioutlet/cart-service/internal/models"
	dapr "github.com/dapr/go-sdk/client"
	"go.uber.org/zap"
)

// ProductClient interface for product service communication
type ProductClient interface {
	GetProduct(ctx context.Context, productID string) (*models.ProductInfo, error)
	GetProducts(ctx context.Context, productIDs []string) ([]models.ProductInfo, error)
}

// productClient implements ProductClient interface using Dapr service invocation
type productClient struct {
	daprClient dapr.Client
	logger     *zap.Logger
}

// NewProductClient creates a new product client using Dapr SDK
func NewProductClient(daprClient dapr.Client, logger *zap.Logger) ProductClient {
	return &productClient{
		daprClient: daprClient,
		logger:     logger,
	}
}

// GetProduct retrieves product information by ID using Dapr service invocation
func (c *productClient) GetProduct(ctx context.Context, productID string) (*models.ProductInfo, error) {
	methodPath := fmt.Sprintf("/api/products/%s", productID)
	
	// Invoke product-service via Dapr
	resp, err := c.daprClient.InvokeMethod(ctx, "product-service", methodPath, "GET")
	if err != nil {
		c.logger.Error("Failed to invoke product service via Dapr", 
			zap.String("productID", productID),
			zap.Error(err))
		return nil, fmt.Errorf("failed to invoke product service: %w", err)
	}

	// Check if product not found (empty response or error structure)
	if len(resp) == 0 {
		return nil, models.ErrProductNotFound
	}

	// Product service returns the product data directly
	var productInfo models.ProductInfo
	if err := json.Unmarshal(resp, &productInfo); err != nil {
		c.logger.Error("Failed to unmarshal product response", 
			zap.String("productID", productID),
			zap.Error(err))
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	// Validate we got a valid product
	if productInfo.ID == "" {
		return nil, models.ErrProductNotFound
	}

	return &productInfo, nil
}

// GetProducts retrieves multiple products by IDs using Dapr service invocation
func (c *productClient) GetProducts(ctx context.Context, productIDs []string) ([]models.ProductInfo, error) {
	methodPath := "/api/products/batch"
	
	requestBody := map[string][]string{
		"productIds": productIDs,
	}

	bodyBytes, err := json.Marshal(requestBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	// Invoke product-service via Dapr with POST method
	content := &dapr.DataContent{
		Data:        bodyBytes,
		ContentType: "application/json",
	}
	resp, err := c.daprClient.InvokeMethodWithContent(ctx, "product-service", methodPath, "POST", content)
	if err != nil {
		c.logger.Error("Failed to invoke product service batch via Dapr", zap.Error(err))
		return nil, fmt.Errorf("failed to invoke product service: %w", err)
	}

	var response struct {
		Success bool                  `json:"success"`
		Data    []models.ProductInfo  `json:"data"`
		Message string                `json:"message"`
	}

	if err := json.Unmarshal(resp, &response); err != nil {
		c.logger.Error("Failed to unmarshal batch products response", zap.Error(err))
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	if !response.Success {
		return nil, fmt.Errorf("product service returned error: %s", response.Message)
	}

	return response.Data, nil
}

File: src/main/java/com/aioutlet/cartservice/client/ProductClient.java
package com.xshopai.cartservice.client;

import com.xshopai.cartservice.model.ProductInfo;
import io.dapr.client.DaprClient;
import io.dapr.client.DaprClientBuilder;
import io.dapr.client.domain.HttpExtension;
import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import jakarta.enterprise.context.ApplicationScoped;
import org.jboss.logging.Logger;

import jakarta.inject.Inject;

@ApplicationScoped
public class ProductClient {
    
    @Inject
    Logger logger;
    
    private DaprClient daprClient;
    
    private static final String PRODUCT_SERVICE_APP_ID = "product-service";
    
    @PostConstruct
    void init() {
        this.daprClient = new DaprClientBuilder().build();
    }
    
    @PreDestroy
    void cleanup() {
        try {
            if (daprClient != null) {
                daprClient.close();
            }
        } catch (Exception e) {
            logger.error("Error closing Dapr client", e);
        }
    }
    
    public ProductInfo getProduct(String productId) {
        try {
            ProductInfo product = daprClient.invokeMethod(
                PRODUCT_SERVICE_APP_ID,
                "/api/products/" + productId,
                null,
                HttpExtension.GET,
                ProductInfo.class
            ).block();
            
            return product;
        } catch (Exception e) {
            logger.errorf("Failed to get product %s: %s", productId, e.getMessage());
            throw new RuntimeException("Failed to get product: " + productId, e);
        }
    }
}
